x/gamm/types/pool.go

```diff
@@ -13,6 +13,34 @@ import (
 type PoolI interface {
 	proto.Message
 
+	GetAddress() sdk.AccAddress
+	String() string
+	GetId() uint64
+	// Todo: may have to generalize this later
+	// Also todo: come up with better name here
+	GetSwapFee(ctx sdk.Context) sdk.Dec
+	// Todo: These may not be needed
+	NumAssets() int
+	IsActive(curBlockTime time.Time) bool
+	// Returns the coins in the pool owned by all LP shareholders
+	GetTotalLpBalances(ctx sdk.Context) sdk.Coins
+	GetTotalShares() sdk.Int
+
+	CalcOutAmtGivenIn(ctx sdk.Context, tokenIn sdk.Coins, tokenOutDenom string, swapFee sdk.Dec) (tokenOut sdk.DecCoin, err error)
```

Comment:
The reason was because of swap fee handling and liquidity during high volatility. One goal I didn't really document anywhere was enabling assymetric swap fees from the underlying AMM. I think the future of AMM's is that they track volatility themselves, and have different swap fees (Spreads) depending on the direction the trade is going during high volatility, and change the amount of effective liquidity. Similar to real market makers!

The alternative is we make the AMM interface do that, but I think the generality may suffer a bit